runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -238,7 +230,19 @@ public void onTimer(
         String timerId, BoundedWindow window, Instant timestamp, TimeDomain timeDomain) {}
 
     @Override
-    public void finishBundle() {}
+    public void finishBundle() {
+      try {
+        // TODO: it would be nice to emit results as they arrive, can thread wait non-blocking?
```

Comment:
I don't think we can safely finish the bundle until all results are received, however there's no reason not to also put the read-and-emit-as-much-as-we-can-from-outputQueue loop into processElement as well.

Reply:
[Not sure if this comment went through.]

This may be due to the fact that there's not enough data to flush these buffered channels until we actually close them. In any case, it shouldn't hurt, and will help with very large bundles.